crates/ra_ide/src/references/rename.rs

```diff
@@ -124,6 +127,116 @@ fn rename_mod(
+    let source_file = sema.parse(position.file_id);
+    let syn = source_file.syntax();
+
+    let fn_def = find_node_at_offset::<ast::FnDef>(syn, position.offset)?;
+    let params = fn_def.param_list()?;
+    if params.self_param().is_some() {
+        return None; // method already has self param
+    }
+    let first_param = params.params().next()?;
+    let mutable = match first_param.ascribed_type() {
+        Some(ast::TypeRef::ReferenceType(rt)) => rt.mut_token().is_some(),
+        _ => return None, // not renaming other types
+    };
+
+    let RangeInfo { range, info: refs } = find_all_refs(db, position, None)?;
+
+    let param_range = first_param.syntax().text_range();
+    let (param_ref, usages): (Vec<Reference>, Vec<Reference>) = refs
+        .into_iter()
+        .partition(|reference| param_range.intersect(reference.file_range.range).is_some());
+
+    if param_ref.is_empty() {
+        return None;
+    }
+
+    let mut edits = usages
+        .into_iter()
+        .map(|reference| source_edit_from_reference(reference, "self"))
+        .collect::<Vec<_>>();
+
+    edits.push(SourceFileEdit {
+        file_id: position.file_id,
+        edit: TextEdit::replace(
+            param_range,
+            String::from(if mutable { "&mut self" } else { "&self" }),
+        ),
+    });
+
+    Some(RangeInfo::new(range, SourceChange::source_file_edits("rename", edits)))
+}
+
+fn source_edit_from_self(
+    syn: &SyntaxNode,
+    position: FilePosition,
+    new_name: &str,
+) -> Option<SourceFileEdit> {
+    fn target_type_name(impl_def: &ast::ImplDef) -> Option<String> {
+        if let Some(ast::TypeRef::PathType(p)) = impl_def.target_type() {
+            return Some(p.path()?.segment()?.name_ref()?.text().to_string());
+        }
+        None
+    }
+
+    let token = syn.token_at_offset(position.offset).find(|t| t.kind() == SyntaxKind::SELF_KW)?;
+    let mut replacement_text = String::from(new_name);
+    let range = if let Some(self_param) = ast::SelfParam::cast(token.parent()) {
+        let impl_def = find_node_at_offset::<ast::ImplDef>(syn, position.offset)?;
+        let type_name = target_type_name(&impl_def)?;
+        replacement_text.push_str(": ");
+        replacement_text.push_str(self_param.mut_token().map_or("&", |_| "&mut "));
+        replacement_text.push_str(type_name.as_str());
+        self_param.syntax().text_range()
+    } else {
+        token.text_range()
+    };
+    Some(SourceFileEdit {
+        file_id: position.file_id,
+        edit: TextEdit::replace(range, replacement_text),
+    })
+}
+
+fn rename_self_to_param(
+    db: &RootDatabase,
+    position: FilePosition,
+    new_name: &str,
+) -> Option<RangeInfo<SourceChange>> {
+    use std::convert::TryInto;
+
+    let sema = Semantics::new(db);
+    let source_file = sema.parse(position.file_id);
+    let syn = source_file.syntax();
+
+    let token = syn.token_at_offset(position.offset).find(|t| t.kind() == SyntaxKind::SELF_KW)?;
```

Comment:
We do this check for self twice, inside and outside the function. Usually, it's better to (waves arms like a windmill) reify the proof that the check has been made as a type. In this case, it means accepting `ast::SelfParam` as a funciton parameter, instead of `FilePosition`.

Reply:
Actually, it's not always `ast::SelfParam` here. `SELF_KW` token can be a part of a path, for example:
```
        fn f(&self) -> i32 {
            self<|>.i
        }
```
So I could not find any fitting type here.

Reply:
Hm, riiight... Yeah, I guess this is fine then (given that the whole rename infra is somewhat aproximate at the moment: #4290)